File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/logging.py
"""Telemetry seam for the attestation library.

The library never talks to a telemetry SDK directly. It emits telemetry through
an injected :class:`AttestationLogger`; by default a no-op logger is installed
so the library is silent until the host wires one up via
:func:`create_attestation_service`.

The free ``track_event`` / ``track_exception`` / ``track_dependency`` functions
are thin delegators to the installed logger; internal modules import those.
Signatures match the Node / .NET logger contract.
"""

from __future__ import annotations

from typing import Any, Mapping, Optional, Protocol, runtime_checkable

TelemetryProps = Optional[Mapping[str, Any]]


@runtime_checkable
class AttestationLogger(Protocol):
    def track_event(
        self,
        name: str,
        properties: TelemetryProps = None,
        measurements: Optional[Mapping[str, float]] = None,
    ) -> None: ...

    def track_exception(self, error: Any, properties: TelemetryProps = None) -> None: ...

    def track_dependency(
        self,
        name: str,
        target: str = "",
        data: str = "",
        duration: float = 0,
        success: bool = True,
        result_code: Any = None,
        dependency_type_name: str = "HTTP",
        properties: TelemetryProps = None,
    ) -> None: ...


class _NullLogger:
    def track_event(self, *args: Any, **kwargs: Any) -> None:  # noqa: D401
        pass

    def track_exception(self, *args: Any, **kwargs: Any) -> None:
        pass

    def track_dependency(self, *args: Any, **kwargs: Any) -> None:
        pass


_logger: AttestationLogger = _NullLogger()


def set_attestation_logger(logger: AttestationLogger) -> None:
    """Install the host logger. Called by ``create_attestation_service``."""
    global _logger
    _logger = logger


def track_event(
    name: str,
    properties: TelemetryProps = None,
    measurements: Optional[Mapping[str, float]] = None,
) -> None:
    _logger.track_event(name, properties, measurements)


def track_exception(error: Any, properties: TelemetryProps = None) -> None:
    _logger.track_exception(error, properties)


def track_dependency(
    name: str,
    target: str = "",
    data: str = "",
    duration: float = 0,
    success: bool = True,
    result_code: Any = None,
    dependency_type_name: str = "HTTP",
    properties: TelemetryProps = None,
) -> None:
    _logger.track_dependency(
        name,
        target,
        data,
        duration,
        success,
        result_code,
        dependency_type_name,
        properties,
    )

File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/logging.ts
/**
 * Telemetry abstraction for the attestation library.
 *
 * The library never depends on a concrete telemetry SDK. The host installs an
 * {@link AttestationLogger} implementation (e.g. App Insights) ONCE at startup
 * via `createAttestationService(config, store, logger)`, and the library emits
 * telemetry through the free functions below, which delegate to the installed
 * logger (defaulting to a no-op so unconfigured/test runs are silent).
 *
 * Call sites keep using `trackEvent` / `trackException` / `trackDependency`
 * exactly as before — only the source of these functions moved from the App
 * Insights module (now app-side) to this framework-agnostic abstraction.
 */

/** Optional structured properties attached to a telemetry item. */
export type TelemetryProps = Record<string, unknown> | undefined;

/** Sink the host provides so the library can emit telemetry without owning an SDK. */
export interface AttestationLogger {
  /** Record a named custom event with optional properties + numeric measurements. */
  trackEvent(name: string, properties?: TelemetryProps, measurements?: Record<string, number>): void;
  /** Record an exception with optional properties. */
  trackException(error: unknown, properties?: TelemetryProps): void;
  /** Record an outbound dependency call (e.g. Play Integrity / revocation HTTP). */
  trackDependency(opts: {
    name: string;
    target?: string;
    data?: string;
    duration: number;
    success: boolean;
    resultCode?: string | number;
    dependencyTypeName?: string;
    properties?: TelemetryProps;
  }): void;
}

const noopLogger: AttestationLogger = {
  trackEvent() {},
  trackException() {},
  trackDependency() {},
};

let logger: AttestationLogger = noopLogger;

/** Install the telemetry logger. Called once by `createAttestationService`. */
export function setAttestationLogger(next: AttestationLogger): void {
  logger = next;
}

export function trackEvent(
  name: string,
  properties?: TelemetryProps,
  measurements?: Record<string, number>,
): void {
  logger.trackEvent(name, properties, measurements);
}

export function trackException(error: unknown, properties?: TelemetryProps): void {
  logger.trackException(error, properties);
}

export function trackDependency(opts: {
  name: string;
  target?: string;
  data?: string;
  duration: number;
  success: boolean;
  resultCode?: string | number;
  dependencyTypeName?: string;
  properties?: TelemetryProps;
}): void {
  logger.trackDependency(opts);
}
